docs/tutorials/lognamespacing.md

```diff
@@ -0,0 +1,229 @@
+        log: &mut LogEvent,
+        legacy_key: Option<LegacyKey<impl ValuePath<'a>>>,
+        metadata_key: impl ValuePath<'a>,
+        value: impl Into<Value>,
+    ) 
+```
+
+#### source_name
+
+The name of the source. This will be eg. `KafkaSourceConfig::NAME`.
+
+#### log
+
+The log event to populate.
+
+#### legacy_key
+
+The field name to populate for the legacy namespace. Pass `None` if
+this field should not be inserted for Legacy. Because there is a 
+possibility that the field might conflict with another field that 
+is already in the event what to do in the case of conflicts must 
+also be specified. `LegacyKey::Overwrite` will overwrite the existing
+value with this value. `LegacyKey::InsertIfEmpty` keeps the original
+value. 
+
+#### metadata_key
+
+The name of the path to insert into the Source metadata when in 
+the Vector namespace. Because there is no chance of conflicting names
+here, this is typically just a hard coded value. eg. `path!("topic")`
+
+#### value
+
+The actual value that is to be inserted into the metadata.
+
+## The event
+
+The main log event should contain only the real log message that the 
+event is representing. 
+
+For the Vector namespace the data should be at the top level and not
+contained in any subfields. For an event that is a single String value -
+typically, in the Legacy namespace this will be inserted in a field 
+called `message`. In the Vector namespace the event will be just this
+String value.
+
+In this case code that creates an event typically looks similar to:
+
+```rust
+    let mut log = match log_namespace {
+        LogNamespace::Vector => LogEvent::from(message),
+        LogNamespace::Legacy => {
+            let mut log = LogEvent::default();
+
+            // Add message
+            log.insert(log_schema().message_key(), message);
+            log
+        }
+    };
+```
+
+Other fields should be inserted into the event like:
+
+```rust
+  log_event.insert(event_path!("path"), value);
+```
+
+## Timestamps
+
+We need to talk about timestamps. A timestamp can represent a number 
+of different things.
+
+- Ingest timestamp - This is the timestamp when the event was received
+  by Vector. This should go in the Vector metadata.
+- Timestamp - This should be any timestamp extracted from the incoming 
+  message. No timestamp 
+  
+It is worth noticing that existing sources have not always been consistent
+with this. Some sources would insert a timestamp that is extracted from the
+event but default to the ingest timestamp if it didn't exist. Others insert 
+the timestamp extracted from the event and don't insert a timestamp at all
+if it didn't exist. Others will always insert the ingest timestamp. To 
+maintain backward compatibility there is a few areas in the code base that
+do some seemingly overly complicated things with timestamps. It is worth 
+bearing this in mind when looking through existing new code.
+
+All new sources should work like the above.
```

Comment:
With regard to new sources, probably worth mentioning (like `amqp`, I think) we didn't allow user configured field names to make things less complicated - new sources should continue to do so.

Reply:
Not sure I understand what you mean?

Reply:
Gotcha

Reply:
It's _sorta_ out of scope for this document, as I think about it today, but I'd expect the times we'd be referring to this doc would be the same time someone would be adding a new source - so it makes _some_ sense (at least to me 😄 )